src/music/midifile.cpp

```diff
@@ -450,8 +794,304 @@ void MidiFile::MoveFrom(MidiFile &other)
+			nexttempoindex++;
+			needtime = true;
+		}
+		/* If a tempo change occurred between two blocks, rather than
+		 * at start of this one, start over with delta time for the block. */
+		if (nexttempo.ticktime < block.ticktime) {
+			goto redoblock;
+		}
+
+		/* Write each block data command */
+		byte *dp = block.data.Begin();
+		while (dp < block.data.End()) {
+			/* Always zero delta time inside blocks */
+			if (needtime) {
+				fwrite(&(bb = 0), 1, 1, f);
+			}
+			needtime = true;
+
+			/* Check message type and write appropriate number of bytes */
+			switch (*dp & 0xF0) {
+			case MIDIST_NOTEOFF:
+			case MIDIST_NOTEON:
+			case MIDIST_POLYPRESS:
+			case MIDIST_CONTROLLER:
+			case MIDIST_PITCHBEND:
+				fwrite(dp, 1, 3, f);
+				dp += 3;
+				continue;
+			case MIDIST_PROGCHG:
+			case MIDIST_CHANPRESS:
+				fwrite(dp, 1, 2, f);
+				dp += 2;
+				continue;
+			}
+
+			/* Sysex needs to measure length and write that as well */
+			if (*dp == MIDIST_SYSEX) {
+				fwrite(dp, 1, 1, f);
+				dp++;
+				byte *sysexend = dp;
+				while (*sysexend++ != MIDIST_ENDSYSEX);
+				ptrdiff_t sysexlen = sysexend - dp;
+				WriteVariableLen(f, sysexlen);
+				fwrite(dp, 1, sysexend - dp, f);
+				dp = sysexend;
+				continue;
+			}
+
+			/* Fail for any other commands */
+			fclose(f);
+			return false;
+		}
+	}
+
+	/* End of track marker */
+	fwrite(&(bb = 0x00), 1, 1, f);
+	fwrite(&(bb = MIDIST_SMF_META), 1, 1, f);
+	fwrite(&(bb = 0x2F), 1, 1, f);
+	fwrite(&(bb = 0x00), 1, 1, f);
+
+	/* Fill out the RIFF block length */
+	size_t trackendpos = ftell(f);
+	fseek(f, tracksizepos, SEEK_SET);
+	uint32 tracksize = trackendpos - tracksizepos - 4;
+	tracksize = TO_BE32(tracksize);
+	fwrite(&tracksize, 4, 1, f);
+
+	fclose(f);
+	return true;
+}
+
+/**
+ * Get the name of a Standard MIDI File for a given song.
+ * For songs already in SMF format, just returns the original.
+ * Otherwise the song is converted and a temporary file written,
+ * and the temporary file's name returned.
+ * @note Caller is responsible for freeing the returned string.
+ * @param song Song definition to query
+ * @return Full filename string, NULL if failed
+ */
+char *MidiFile::GetSMFFile(const MusicSongInfo &song)
+{
+	if (song.filetype == MTT_STANDARDMIDI) {
+		return stredup(song.filename);
+	}
+
+	if (song.filetype != MTT_MPSMIDI) return NULL;
+
+	const char *lastpathsep = strrchr(song.filename, PATHSEPCHAR);
+	if (lastpathsep == NULL) {
+		lastpathsep = song.filename;
+	}
+
+	char basename[MAX_PATH];
+	{
+		/* Remove all '.' characters from filename */
+		char *wp = basename;
+		for (const char *rp = lastpathsep + 1; *rp != '\0'; rp++) {
+			if (*rp != '.') *wp++ = *rp;
+		}
+		*wp++ = '\0';
+	}
+
+	char tempdirname[MAX_PATH];
+	FioGetFullPath(tempdirname, lastof(tempdirname), Searchpath::SP_AUTODOWNLOAD_DIR, Subdirectory::BASESET_DIR, basename);
+	if (AppendPathSeparator(tempdirname, lastof(tempdirname)) == false) return NULL;
+	FioCreateDirectory(tempdirname);
```

Comment:
does this really need its own directory? midi's are already unpacked and they didn't require a new directory previously...

Reply:
These are specifically files that get decoded ("unpacked") from a CAT file, to support music drivers that only work by loading named files from disk.

Reply:
Actually I would rather have just written to a system temp dir, but don't want to litter it with a new file every time song changes, and just getting the temp dir with a custom filename appended seems to be a lot of work to do in a portable manner.